File: tests/wasm/test_kst.js
/**
 * WASM binding tests for KST indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('KST default params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.kst(close, 10, 10, 10, 15, 10, 15, 20, 30, 9);
    assert(result.values, 'Result should have values property');
    assert(result.rows, 'Result should have rows property');
    assert(result.cols, 'Result should have cols property');
    assert.strictEqual(result.rows, 2);
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.values.length, 2 * close.length);
    
    
    const line = result.values.slice(0, result.cols);
    const signal = result.values.slice(result.cols, 2 * result.cols);
    assert.strictEqual(line.length, close.length);
    assert.strictEqual(signal.length, close.length);
});

test('KST accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.kst;
    
    const result = wasm.kst(
        close,
        expected.defaultParams.sma_period1,
        expected.defaultParams.sma_period2,
        expected.defaultParams.sma_period3,
        expected.defaultParams.sma_period4,
        expected.defaultParams.roc_period1,
        expected.defaultParams.roc_period2,
        expected.defaultParams.roc_period3,
        expected.defaultParams.roc_period4,
        expected.defaultParams.signal_period
    );
    
    
    const line = result.values.slice(0, result.cols);
    const signal = result.values.slice(result.cols, 2 * result.cols);
    
    assert.strictEqual(line.length, close.length);
    assert.strictEqual(signal.length, close.length);
    
    
    const last5Line = line.slice(-5);
    const last5Signal = signal.slice(-5);
    assertArrayClose(
        last5Line,
        expected.last5Values.line,
        1e-1,  
        "KST line last 5 values mismatch"
    );
    assertArrayClose(
        last5Signal,
        expected.last5Values.signal,
        1e-1,  
        "KST signal last 5 values mismatch"
    );
    
    
    
});

test('KST all NaN values', () => {
    
    const inputData = new Float64Array(10).fill(NaN);
    
    assert.throws(() => {
        wasm.kst(inputData, 10, 10, 10, 15, 10, 15, 20, 30, 9);
    }, /All values are NaN/);
});

test('KST zero periods', () => {
    
    const inputData = new Float64Array(50).fill(0).map((_, i) => 10.0 + i);
    
    
    assert.throws(() => {
        wasm.kst(inputData, 0, 10, 10, 15, 10, 15, 20, 30, 9);
    }, /period|Period/);
    
    
    assert.throws(() => {
        wasm.kst(inputData, 10, 10, 10, 15, 0, 15, 20, 30, 9);
    }, /period|Period/);
    
    
    assert.throws(() => {
        wasm.kst(inputData, 10, 10, 10, 15, 10, 15, 20, 30, 0);
    }, /period|Period/);
});

test('KST insufficient data', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0, 40.0, 50.0]);
    
    assert.throws(() => {
        wasm.kst(dataSmall, 10, 10, 10, 15, 10, 15, 20, 30, 9);
    }, /Not enough|insufficient|Invalid/);
});

test('KST with NaN prefix', () => {
    
    const inputData = new Float64Array(100);
    inputData.fill(NaN, 0, 20);
    for (let i = 20; i < 100; i++) {
        inputData[i] = 100.0 + Math.sin(i * 0.1) * 10.0;
    }
    
    const result = wasm.kst(inputData, 5, 5, 5, 10, 5, 10, 15, 20, 5);
    
    
    const line = result.values.slice(0, result.cols);
    const signal = result.values.slice(result.cols, 2 * result.cols);
    
    assert.strictEqual(line.length, inputData.length);
    assert.strictEqual(signal.length, inputData.length);
    
    
    
    
    
    
    const warmup = 49;
    for (let i = 0; i < warmup; i++) {
        assert(isNaN(line[i]), `Line[${i}] should be NaN during warmup`);
        assert(isNaN(signal[i]), `Signal[${i}] should be NaN during warmup`);
    }
    
    
    for (let i = warmup + 10; i < line.length; i++) {
        assert(!isNaN(line[i]), `Line[${i}] should not be NaN after warmup`);
        assert(!isNaN(signal[i]), `Signal[${i}] should not be NaN after warmup`);
    }
});

test('KST fast/unsafe API', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const inPtr = wasm.kst_alloc(len);
    const lineOutPtr = wasm.kst_alloc(len);
    const signalOutPtr = wasm.kst_alloc(len);
    
    try {
        
        const inArray = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);
        
        
        inArray.set(close);
        
        
        wasm.kst_into(
            inPtr, lineOutPtr, signalOutPtr, len,
            10, 10, 10, 15, 10, 15, 20, 30, 9
        );
        
        
        const memBuffer1 = wasm.__wasm.memory.buffer;
        const lineResult = new Float64Array(memBuffer1, lineOutPtr, len);
        const signalResult = new Float64Array(memBuffer1, signalOutPtr, len);
        
        
        const safeResult = wasm.kst(close, 10, 10, 10, 15, 10, 15, 20, 30, 9);
        const safeLine = safeResult.values.slice(0, safeResult.cols);
        const safeSignal = safeResult.values.slice(safeResult.cols, 2 * safeResult.cols);
        
        
        const memBuffer2 = wasm.__wasm.memory.buffer;
        const lineResultFinal = memBuffer1 === memBuffer2 ? lineResult : new Float64Array(memBuffer2, lineOutPtr, len);
        const signalResultFinal = memBuffer1 === memBuffer2 ? signalResult : new Float64Array(memBuffer2, signalOutPtr, len);
        
        assertArrayClose(
            Array.from(lineResultFinal),
            safeLine,
            1e-10,
            "Fast API line should match safe API"
        );
        assertArrayClose(
            Array.from(signalResultFinal),
            safeSignal,
            1e-10,
            "Fast API signal should match safe API"
        );
    } finally {
        
        wasm.kst_free(inPtr, len);
        wasm.kst_free(lineOutPtr, len);
        wasm.kst_free(signalOutPtr, len);
    }
});

test('KST fast API with aliasing', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const ptr = wasm.kst_alloc(len);
    const signalPtr = wasm.kst_alloc(len);
    
    try {
        
        const array = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        
        
        array.set(close);
        
        
        wasm.kst_into(
            ptr, ptr, signalPtr, len,
            10, 10, 10, 15, 10, 15, 20, 30, 9
        );
        
        
        const lineResult = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        const signalResult = new Float64Array(wasm.__wasm.memory.buffer, signalPtr, len);
        
        
        const safeResult = wasm.kst(close, 10, 10, 10, 15, 10, 15, 20, 30, 9);
        const safeLine = safeResult.values.slice(0, safeResult.cols);
        const safeSignal = safeResult.values.slice(safeResult.cols, 2 * safeResult.cols);
        
        assertArrayClose(
            Array.from(lineResult),
            safeLine,
            1e-10,
            "Aliased fast API line should match safe API"
        );
        assertArrayClose(
            Array.from(signalResult),
            safeSignal,
            1e-10,
            "Aliased fast API signal should match safe API"
        );
    } finally {
        
        wasm.kst_free(ptr, len);
        wasm.kst_free(signalPtr, len);
    }
});

test('KST reinput', () => {
    
    const close = new Float64Array(testData.close.slice(0, 500));
    
    
    const firstResult = wasm.kst(close, 10, 10, 10, 15, 10, 15, 20, 30, 9);
    const firstLine = firstResult.values.slice(0, firstResult.cols);
    const firstSignal = firstResult.values.slice(firstResult.cols, 2 * firstResult.cols);
    
    
    const secondResult = wasm.kst(firstLine, 10, 10, 10, 15, 10, 15, 20, 30, 9);
    const secondLine = secondResult.values.slice(0, secondResult.cols);
    const secondSignal = secondResult.values.slice(secondResult.cols, 2 * secondResult.cols);
    
    assert.strictEqual(secondLine.length, firstLine.length);
    
    
    const warmupFirst = 44;
    
    
    
    const warmupSecond = 88;
    
    
    assert(isNaN(firstLine[0]), 'First line should have NaN at start');
    
    
    assert(isNaN(secondLine[0]), 'Second line should have NaN at start');
    assert(isNaN(secondLine[43]), 'Second line should have NaN during first warmup');
    
    
    if (secondLine.length > warmupSecond + 10) {
        assert(!isNaN(secondLine[warmupSecond + 10]), 'Should have valid value after cascaded warmup');
    }
});

test('KST warmup period', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    const result = wasm.kst(close, 10, 10, 10, 15, 10, 15, 20, 30, 9);
    const line = result.values.slice(0, result.cols);
    const signal = result.values.slice(result.cols, 2 * result.cols);
    
    
    
    const expectedWarmup = 44;
    
    
    for (let i = 0; i < expectedWarmup; i++) {
        assert(isNaN(line[i]), `Expected NaN at index ${i} during warmup`);
    }
    
    
    for (let i = expectedWarmup; i < Math.min(expectedWarmup + 5, line.length); i++) {
        assert(!isNaN(line[i]), `Expected valid value at index ${i} after warmup`);
    }
    
    
    const signalWarmup = expectedWarmup + 9 - 1;
    for (let i = 0; i < signalWarmup; i++) {
        assert(isNaN(signal[i]), `Expected NaN in signal at index ${i} during warmup`);
    }
});

test('KST empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.kst(empty, 10, 10, 10, 15, 10, 15, 20, 30, 9);
    }, /empty|Empty/, 'Should fail with empty input');
});

test('KST batch calculation', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    const config = {
        sma_period1: [10, 10, 0],  
        sma_period2: [10, 10, 0],
        sma_period3: [10, 10, 0],
        sma_period4: [15, 15, 0],
        roc_period1: [10, 12, 2],  
        roc_period2: [15, 15, 0],
        roc_period3: [20, 20, 0],
        roc_period4: [30, 30, 0],
        signal_period: [9, 11, 2]  
    };
    
    const result = wasm.kst_batch(close, config);
    
    assert(result.values, 'Batch result should have values property');
    assert(result.combos, 'Batch result should have combos property');
    assert(result.rows, 'Batch result should have rows property');
    assert(result.cols, 'Batch result should have cols property');
    
    
    const numCombos = 2 * 2; 
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.combos.length, numCombos);
    assert.strictEqual(result.rows, numCombos * 2); 
    assert.strictEqual(result.values.length, result.rows * result.cols);
    
    
    const firstCombo = result.combos[0];
    const firstLineRow = result.values.slice(0, result.cols);
    const firstSignalRow = result.values.slice(numCombos * result.cols, (numCombos + 1) * result.cols);
    
    
    const singleResult = wasm.kst(
        close,
        firstCombo.sma_period1,
        firstCombo.sma_period2,
        firstCombo.sma_period3,
        firstCombo.sma_period4,
        firstCombo.roc_period1,
        firstCombo.roc_period2,
        firstCombo.roc_period3,
        firstCombo.roc_period4,
        firstCombo.signal_period
    );
    
    const singleLine = singleResult.values.slice(0, singleResult.cols);
    const singleSignal = singleResult.values.slice(singleResult.cols, 2 * singleResult.cols);
    
    assertArrayClose(
        firstLineRow,
        singleLine,
        1e-10,
        "Batch first row line should match single calculation"
    );
    assertArrayClose(
        firstSignalRow,
        singleSignal,
        1e-10,
        "Batch first row signal should match single calculation"
    );
});

console.log('All KST WASM tests passed! 🎉');

File: tests/python/test_kst.py
"""
Python binding tests for KST indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestKst:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_kst_default_params(self, test_data):
        """Test KST with default parameters - mirrors check_kst_default_params"""
        close = test_data['close']


        line, signal = ta_indicators.kst(
            close,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )
        assert len(line) == len(close)
        assert len(signal) == len(close)

    def test_kst_accuracy(self, test_data):
        """Test KST matches expected values from Rust tests - mirrors check_kst_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['kst']


        line, signal = ta_indicators.kst(
            close,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )

        assert len(line) == len(close)
        assert len(signal) == len(close)


        assert_close(
            line[-5:],
            expected['last_5_line'],
            rtol=0,
            atol=1e-1,
            msg="KST line last 5 values mismatch"
        )

        assert_close(
            signal[-5:],
            expected['last_5_signal'],
            rtol=0,
            atol=1e-1,
            msg="KST signal last 5 values mismatch"
        )


        compare_with_rust('kst_line', line)
        compare_with_rust('kst_signal', signal)

    def test_kst_partial_params(self, test_data):
        """Test KST with partial parameters"""
        close = test_data['close']


        line, signal = ta_indicators.kst(
            close,
            sma_period1=12, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=12, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=10
        )
        assert len(line) == len(close)
        assert len(signal) == len(close)

    def test_kst_zero_period(self):
        """Test KST fails with zero period"""
        data = np.array([10.0, 20.0, 30.0] * 20)

        with pytest.raises(ValueError):
            ta_indicators.kst(data, sma_period1=0)

        with pytest.raises(ValueError):
            ta_indicators.kst(data, roc_period1=0)

        with pytest.raises(ValueError):
            ta_indicators.kst(data, signal_period=0)

    def test_kst_period_exceeds_data(self):
        """Test KST fails when period exceeds data length"""
        small_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError):
            ta_indicators.kst(small_data, roc_period4=50)

    def test_kst_nan_handling(self, test_data):
        """Test KST handles NaN values correctly - mirrors check_kst_nan_handling"""
        all_nan = np.full(10, np.nan)

        with pytest.raises(ValueError):
            ta_indicators.kst(all_nan)

    def test_kst_all_params(self, test_data):
        """Test KST with all parameters specified"""
        close = test_data['close']

        line, signal = ta_indicators.kst(
            close,
            sma_period1=10,
            sma_period2=10,
            sma_period3=10,
            sma_period4=15,
            roc_period1=10,
            roc_period2=15,
            roc_period3=20,
            roc_period4=30,
            signal_period=9
        )

        assert len(line) == len(close)
        assert len(signal) == len(close)

    def test_kst_kernel_selection(self, test_data):
        """Test KST with different kernel selections"""
        close = test_data['close']


        line_auto, signal_auto = ta_indicators.kst(
            close,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9,
            kernel=None
        )
        line_scalar, signal_scalar = ta_indicators.kst(
            close,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9,
            kernel='scalar'
        )

        assert len(line_auto) == len(close)
        assert len(line_scalar) == len(close)


        assert_close(line_auto, line_scalar, rtol=1e-10)
        assert_close(signal_auto, signal_scalar, rtol=1e-10)

    def test_kst_streaming(self, test_data):
        """Test KST streaming functionality"""
        close = np.array(test_data['close'])






        first_valid = next((i for i, v in enumerate(close) if not np.isnan(v)), 0)
        valid_close = close[first_valid:]


        stream = ta_indicators.KstStream(10, 10, 10, 15, 10, 15, 20, 30, 9)


        stream_results = []
        for price in valid_close:
            if not np.isnan(price):
                result = stream.update(price)
                stream_results.append(result)


        line_results = []
        signal_results = []
        for r in stream_results:
            if r is None:
                line_results.append(np.nan)
                signal_results.append(np.nan)
            else:
                line_results.append(r[0])
                signal_results.append(r[1])


        batch_line, batch_signal = ta_indicators.kst(
            valid_close,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )


        assert_close(line_results, batch_line, rtol=1e-10)











        valid_lines = sum(1 for v in line_results if not np.isnan(v))
        assert valid_lines > 0, "Stream should produce some valid line values"

    def test_kst_batch_single_parameter_set(self, test_data):
        """Test batch processing with single parameter combination"""
        close = test_data['close']


        batch_result = ta_indicators.kst_batch(
            close,
            sma1_range=(10, 10, 0),
            sma2_range=(10, 10, 0),
            sma3_range=(10, 10, 0),
            sma4_range=(15, 15, 0),
            roc1_range=(10, 10, 0),
            roc2_range=(15, 15, 0),
            roc3_range=(20, 20, 0),
            roc4_range=(30, 30, 0),
            sig_range=(9, 9, 0)
        )


        assert 'line' in batch_result
        assert 'signal' in batch_result
        assert 'sma1' in batch_result
        assert 'sma2' in batch_result
        assert 'sma3' in batch_result
        assert 'sma4' in batch_result
        assert 'roc1' in batch_result
        assert 'roc2' in batch_result
        assert 'roc3' in batch_result
        assert 'roc4' in batch_result
        assert 'sig' in batch_result


        assert batch_result['line'].shape == (1, len(close))
        assert batch_result['signal'].shape == (1, len(close))


        single_line, single_signal = ta_indicators.kst(
            close,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )
        assert_close(batch_result['line'][0], single_line, rtol=1e-10)
        assert_close(batch_result['signal'][0], single_signal, rtol=1e-10)

    def test_kst_batch_multiple_parameters(self, test_data):
        """Test batch processing with multiple parameter combinations"""
        close = test_data['close'][:200]


        batch_result = ta_indicators.kst_batch(
            close,
            sma1_range=(10, 10, 0),
            sma2_range=(10, 10, 0),
            sma3_range=(10, 10, 0),
            sma4_range=(15, 15, 0),
            roc1_range=(10, 12, 2),
            roc2_range=(15, 15, 0),
            roc3_range=(20, 20, 0),
            roc4_range=(30, 30, 0),
            sig_range=(8, 10, 2)
        )


        assert batch_result['line'].shape[0] == 4
        assert batch_result['line'].shape[1] == len(close)
        assert batch_result['signal'].shape[0] == 4
        assert batch_result['signal'].shape[1] == len(close)


        assert len(batch_result['roc1']) == 4
        assert len(batch_result['sig']) == 4

    def test_kst_batch_custom_ranges(self, test_data):
        """Test batch processing with custom parameter ranges"""
        close = test_data['close'][:100]

        batch_result = ta_indicators.kst_batch(
            close,
            sma1_range=(8, 12, 2),
            sma2_range=(10, 10, 0),
            sma3_range=(10, 10, 0),
            sma4_range=(15, 15, 0),
            roc1_range=(10, 10, 0),
            roc2_range=(15, 15, 0),
            roc3_range=(20, 20, 0),
            roc4_range=(25, 30, 5),
            sig_range=(9, 9, 0)
        )


        assert batch_result['line'].shape[0] == 6
        assert batch_result['signal'].shape[0] == 6

    def test_kst_edge_cases(self):
        """Test KST with edge case inputs"""



        min_data = np.random.rand(53)
        line, signal = ta_indicators.kst(
            min_data,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )
        assert len(line) == len(min_data)
        assert len(signal) == len(min_data)



        assert np.isnan(line[0])
        assert np.isnan(signal[0])


        assert not np.isnan(line[-1])
        assert not np.isnan(signal[-1])

    def test_kst_batch_kernel_selection(self, test_data):
        """Test KST batch with different kernels"""
        close = test_data['close'][:100]


        result_auto = ta_indicators.kst_batch(
            close,
            sma1_range=(10, 10, 0),
            sma2_range=(10, 10, 0),
            sma3_range=(10, 10, 0),
            sma4_range=(15, 15, 0),
            roc1_range=(10, 10, 0),
            roc2_range=(15, 15, 0),
            roc3_range=(20, 20, 0),
            roc4_range=(30, 30, 0),
            sig_range=(9, 9, 0),
            kernel=None
        )
        result_scalar = ta_indicators.kst_batch(
            close,
            sma1_range=(10, 10, 0),
            sma2_range=(10, 10, 0),
            sma3_range=(10, 10, 0),
            sma4_range=(15, 15, 0),
            roc1_range=(10, 10, 0),
            roc2_range=(15, 15, 0),
            roc3_range=(20, 20, 0),
            roc4_range=(30, 30, 0),
            sig_range=(9, 9, 0),
            kernel='scalar'
        )


        assert_close(result_auto['line'], result_scalar['line'], rtol=1e-10)
        assert_close(result_auto['signal'], result_scalar['signal'], rtol=1e-10)

    def test_kst_reinput(self, test_data):
        """Test KST applied twice (re-input) - similar to ALMA reinput test"""
        close = test_data['close'][:500]


        first_line, first_signal = ta_indicators.kst(
            close,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )
        assert len(first_line) == len(close)


        second_line, second_signal = ta_indicators.kst(
            first_line,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )
        assert len(second_line) == len(first_line)




        warmup_first = 44
        warmup_second = warmup_first + 44


        assert np.isnan(first_line[0])
        assert np.isnan(second_line[warmup_second - 1])


        if len(second_line) > warmup_second + 10:
            assert not np.isnan(second_line[warmup_second + 10])

    def test_kst_nan_prefix(self, test_data):
        """Test KST with NaN values at start of data"""
        close = test_data['close'][:200]


        close_with_nan = np.array(close)
        close_with_nan[:20] = np.nan

        line, signal = ta_indicators.kst(
            close_with_nan,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )

        assert len(line) == len(close_with_nan)
        assert len(signal) == len(close_with_nan)



        expected_warmup = 44 + 20


        assert np.isnan(line[expected_warmup - 1])


        if len(line) > expected_warmup + 10:
            assert not np.isnan(line[expected_warmup + 10])

    def test_kst_warmup_period(self, test_data):
        """Test KST warmup period calculation"""
        close = test_data['close'][:100]


        line, signal = ta_indicators.kst(
            close,
            sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
            roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
            signal_period=9
        )



        expected_warmup = 44


        for i in range(expected_warmup):
            assert np.isnan(line[i]), f"Expected NaN at index {i} during warmup"


        for i in range(expected_warmup, min(expected_warmup + 5, len(line))):
            assert not np.isnan(line[i]), f"Expected valid value at index {i} after warmup"


        signal_warmup = expected_warmup + 9 - 1
        for i in range(signal_warmup):
            assert np.isnan(signal[i]), f"Expected NaN in signal at index {i} during warmup"

    def test_kst_invalid_parameters(self):
        """Test KST with invalid parameters"""
        data = np.random.rand(100)


        with pytest.raises(ValueError):
            ta_indicators.kst(
                data,
                sma_period1=0, sma_period2=10, sma_period3=10, sma_period4=15,
                roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
                signal_period=9
            )

        with pytest.raises(ValueError):
            ta_indicators.kst(
                data,
                sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
                roc_period1=0, roc_period2=15, roc_period3=20, roc_period4=30,
                signal_period=9
            )

        with pytest.raises(ValueError):
            ta_indicators.kst(
                data,
                sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
                roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
                signal_period=0
            )

    def test_kst_empty_input(self):
        """Test KST with empty input data"""
        empty = np.array([])

        with pytest.raises(ValueError, match="empty|Empty"):
            ta_indicators.kst(
                empty,
                sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
                roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
                signal_period=9
            )

    def test_kst_insufficient_data(self):
        """Test KST with insufficient data for calculation"""

        small_data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])

        with pytest.raises(ValueError):
            ta_indicators.kst(
                small_data,
                sma_period1=10, sma_period2=10, sma_period3=10, sma_period4=15,
                roc_period1=10, roc_period2=15, roc_period3=20, roc_period4=30,
                signal_period=9
            )